### PR TITLE
Use multiple tasklets

### PR DESCRIPTION
**Pull request by johnramsden**
# Summary

Modify code to use multiple tasklets.

* [x] Split data between tasklets

### CODE REVIEW

**Comment by johnramsden** [comment]
@jnider this is the issue I mentioned.

I have not split up the data between the tasklets yet, but there are only reads from global variables, as a result I would just expect double the results if I ran two tasklets.

I made a few minor modifications to ensure data was set up before the tasklets were run via a handshake and notify:

https://github.com/UBC-ECE-Sasha/PIM-HDC/blob/8a6abdf653326adb4ed2cc19e7f378d97798f99d/PIM_HDC/src/dpu/dpu_task.c#L145-L155

Other than that it's basically the same code. I did make the data a global so it is not duplicated between the tasklets, this is just temporary and eventually I plan on allocating buffers just the size I need for each tasklet.

I added the check you mentioned:

https://github.com/UBC-ECE-Sasha/PIM-HDC/blob/8a6abdf653326adb4ed2cc19e7f378d97798f99d/PIM_HDC/src/dpu/hdc/aux_functions.c#L63-L70

And it does appear the addresses are off:

They're only off for tasklet 1. And I'm getting garbage data in `ix`:

```
compute_N_gram: 1: &channel_iM[-1691009797][253] = 105e7b50, &channel_AM[0][253] = ab5c
```

I've also added and assert, which checks the data `input[]` is greater than zero, and assert always passes. Therefore this number should never be negative

https://github.com/UBC-ECE-Sasha/PIM-HDC/blob/8a6abdf653326adb4ed2cc19e7f378d97798f99d/PIM_HDC/src/dpu/dpu_task.c#L81-L82

Do you have any recommendations? The code still runs fine with one tasklet.

**Comment by johnramsden** [comment]
@jnider multi-tasklets are working now (up to 7 due to stack space). Some additional comments, and code cleanup are necessary but shouldn't be a blocker for the merge.

I had to make a few changes:

* It's not feasible to dynamically allocate all my input variables since I run out of heap space, so I have defined a couple of variables that are the maximum size that should be used. This allows the use of static buffers in key areas that would not fit on the Heap when dynamically allocated. As a result if the input data has larger size two variables the compile time variable sizes will need to be increased. https://github.com/UBC-ECE-Sasha/PIM-HDC/blob/c6ee266b2dac41b712b844dbff11069a42beb3e9/PIM_HDC/Makefile#L19-L20
* I was able to greatly decrease the amount of WRAM needed by a function by noticing only a small subset of the array with used in each loop so we can use a smaller buffer that is the size of the subset https://github.com/UBC-ECE-Sasha/PIM-HDC/pull/20/commits/62980094d766bcccfb2e9e0ef9c09de559950082 .
* For some strange reason I need a lock around a read, or else I get invalid data in several of my outputs. This invalid data is variable and its location and value and changes as I add print statements. What's strange is the two variables being read are never modified, and the variable being written to is local. I am puzzled by this issue. https://github.com/UBC-ECE-Sasha/PIM-HDC/blob/62980094d766bcccfb2e9e0ef9c09de559950082/PIM_HDC/src/dpu/hdc/aux_functions.c#L76-L79

File: PIM_HDC/src/dpu/dpu_task.c [hunk 1]
```diff
@@ -82,15 +83,16 @@ static void alloc_chunks(uint32_t **buf, __mram_ptr uint8_t * mram_ptr, uint32_t
...
- * @param[out] read_buf    Buffers filled with sample data.
  * @return                 @p ENOMEM on failure. Zero on success.
  */
-static int alloc_buffers(int32_t **read_buf) {
+static int alloc_buffers(uint32_t out_size) {
+    output = mem_alloc(out_size);
+
     uint32_t transfer_size = ALIGN(buffer_channel_aligned_size, 8);
-    *read_buf = mem_alloc(channels * transfer_size);
+    read_buf = mem_alloc(channels * transfer_size);
```
Comment (jnider) [inline PIM_HDC/src/dpu/dpu_task.c:97]: Now I understand the problem a bit better. You have put in a lot of effort with the mem_alloc() and handshake, etc. to make these buffers completely dynamic. Maybe there is an easier way that will simplify the code, and debugging. What is the largest number of channels that you can have? What is the largest transfer size? Is it possible to declare a static array of the largest size? This will also help you manage the memory (in your mind) to be able to squeeze in more threads.

File: PIM_HDC/src/dpu/hdc/aux_functions.c [hunk 2]
```diff
@@ -1,10 +1,13 @@
 #include <string.h>
 #include <built_ins.h>
 #include <alloc.h>
+#include <mutex.h>
 
 #include "aux_functions.h"
 #include "cycle_counter.h"
+#include "common.h"
 
+MUTEX_INIT(chHV_mutex);
```
Comment (jnider) [inline PIM_HDC/src/dpu/hdc/aux_functions.c:11]: What is this mutex protecting? The chHV array? Does that mean two tasklets can update the same values in the array at the same time? If so, it's probably better to fix the conflicts than use a mutex.

File: PIM_HDC/src/main.c [hunk 3]
```diff
@@ -374,9 +374,8 @@ int main(int argc, char **argv) {
     }
 
     uint32_t buffer_size = (sizeof(int32_t) * number_of_input_samples * channels);
-    int32_t *data_set;
-
-    if ((data_set = malloc(buffer_size)) == NULL) {
+    int32_t *data_set = malloc(buffer_size);
```
Comment (jnider) [inline PIM_HDC/src/main.c:420]: Both are correct. Just curious - why did you change the syntax?

Comment (johnramsden) [reply PIM_HDC/src/main.c:420]: I guess I just thought it looked cleaner.

File: PIM_HDC/src/main.c [hunk 4]
```diff
@@ -23,6 +23,46 @@
 
 #define DPU_PROGRAM "src/dpu/hdc.dpu"
 
+static dpu_input_data setup_dpu_data(uint32_t buffer_channel_length) {
```
Comment (jnider) [inline PIM_HDC/src/main.c:26]: This looks like a better idea!

File: PIM_HDC/src/dpu/hdc/aux_functions.c [hunk 5]
```diff
@@ -63,35 +64,36 @@ void hamming_dist(uint32_t q[bit_dim + 1], uint32_t *aM, int sims[CLASSES]){
  */
 void compute_N_gram(int32_t input[channels], uint32_t *channel_iM, uint32_t *channel_AM, uint32_t query[bit_dim + 1]) {
 
-    // Pseudo-2d array: uint32_t chHV[channels + 1][bit_dim + 1];
+    // Pseudo-2d array:
 
-    // Not ideal...
-    mutex_lock(chHV_mutex);
+    uint32_t chHV[MAX_CHANNELS + 1] = {0};
```
Comment (jnider) [inline PIM_HDC/src/dpu/hdc/aux_functions.c:69]: Good - I see the area protected by the mutex is much smaller in this version. Maybe we can get it to disappear completely.

File: PIM_HDC/src/dpu/hdc/aux_functions.c (hunk 2 above)
Comment (johnramsden) [reply PIM_HDC/src/dpu/hdc/aux_functions.c:11]: Technically they shouldn't ever conflict since the data should be written once and then only ever read when multiple task lists are active (after the barrier). If I don't have the lock I get incorrect data on one or two results. I'm still working on resolving this issue. Perhaps using static buffers will help, though I'm not sure.

**Comment by johnramsden** [comment]
@jnider so I attempted static buffers https://github.com/UBC-ECE-Sasha/PIM-HDC/pull/20/commits/5946b3a1f68d8c9882d6a9b2c0cf1f9fe361122b , unfortunately it did not allow me to use more tasklets, I run out of stack space. If I'm using more DPUs I _may_ however be able to use more tasklets.

A bonus is I don't need to copy into MRAM and then into WRAM I can just go straight into WRAM. Additionally I was able to get rid of the mutex, meaning I assume there is some sort of error occurring when heap space is used and a variable is read from by multiple tasklets.

I kept the "input data" in a dynamic buffer since it's the _most_ dynamic dataset, that is potentially changing more often. It can also be drastically different size depending on how many DPUs we use. I think this is okay but let me know if you think I should also stick it in a static buffer.

File: PIM_HDC/src/dpu/hdc/aux_functions.c (hunk 2 above)
Comment (johnramsden) [reply PIM_HDC/src/dpu/hdc/aux_functions.c]: Resolved in https://github.com/UBC-ECE-Sasha/PIM-HDC/commit/5946b3a1f68d8c9882d6a9b2c0cf1f9fe361122b
